staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -178,19 +185,64 @@ func (q *Type) Done(item interface{}) {
 	if q.dirty.has(item) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
+	} else if q.processing.len() == 0 {
+		q.cond.Signal()
```

Comment:
Trying to reason through this.  In a typical usage we have multiple go routines doing `.Get` which has a `.wait` today.  The `.signal` from an Add and here generally ensure that when an item is added to the queue we have a matching `.signal` for every `.wait`.  This is leveraged in `.Get` so that a signal with an empty queue indicates that there is no more work to do.

This new signal appears to used to indicate that there is no processing.  Now the `.Get` calls will continue returning items until the queue is empty, then they `.wait`.

If we have two workers doing Gets and one call to Shudown and item/A  I think we can a situation like....
1. Get/1 receives item/A
2. Get/2 enters and waits
3. Shutdown is called and enters a wait
4. we now have two waiters
5. Done for item/A is called and signals
6. Get/2 is signaled from Done and exits as shutting down
7. Get/1 enters and exits as shutting down
8. Shutdown is never signaled

Is that a flow that can happen in this scenario?

Do you actually need the if-block in here to test for shutdown, number in processing, the number in queue, the number in dirty and broadcast?

If I'm correct, I'd like to see a test trying to force this race and fix it.  If I'm incorrect, but the story was plausible to either @alexanderConstantinescu or @lavalamp I think we also need this test.  If I'm incorrect, but the story was not plausible, please help me see the flaw.

Reply:
> the story was plausible to either @alexanderConstantinescu or @lavalamp I think we also need this test.

The story is plausible in my opinion. I will write a test for it and verify. If this is true, I suspect doing 

```go
else if q.processing.len() == 0 {		
        q.cond.Broadcast()
}
```

should fix it, as that would lead to:

5. Done for item/A is called and signals
6. Get/2 and Shutdown is signaled from Done and exits as shutting down
7. Get/1 enters and exits as shutting down

Reply:
After having thought about this some more and written a small test to verify: I've realized that the scenario presented is not completely correct. The following is the actual chain of events:

1. Get/1 receives item/A
2. Get/2 enters and waits
3. Shutdown is called, broadcasts `shuttingDown` and enters a wait
4. Get/2 is signaled from `shuttingDown` and exits as shutting down
4. we now have one waiter
5. Done for item/A is called and signals
6. Shutdown is signaled from Done and exists

Given that `ShutDown` always broadcasts to all `Get`s that they should finish waiting (so, either process one of the remaining items on the queue or just exit if there are none), and also broadcasts to all `Add`s to stop accepting items: we should be fine.